### ISSUE TITLE
Alpha Magnet

### ISSUE THREAD

**Issue by ax3l** [BLAST-ImpactX/impactx#370]
Based on an email question:
We could include a model for Alpha Magnet elements.

This _might_ need an magnetostatic model on top of the electrostatic model (to be confirmed).
For t-based modeling, we have both EM & ES in WarpX.

References:

- P. Guimbal et al., “Status of the ELSA-2 Project,” in Proc. EPAC’02, Paris, France, Jun. 2002, paper TUPRI066, pp.1768-1770.
  https://accelconf.web.cern.ch/e02/PAPERS/TUPRI066.pdf
- TraceWin documentation, page 155
  [https://irfu.cea.fr/dacm/logiciels/codesdacm/tracewin/tracewin.pdf](https://irfu.cea.fr/dacm/logiciels/codesdacm/tracewin/tracewin.pdf)

**Comment by KroneBonn** [comment]
Just wanted to support this idea. For me personally this would be a great addition and something which would make life a lot easier especially compared to other codes.

**Comment by ax3l** [comment]
Merged from #1272:

> Hello,
> 
> In [#370](https://github.com/BLAST-ImpactX/impactx/issues/370), ax3l mentioned the potential to implement alpha magnets into ImpactX. I would like to follow up on this and ask if there are any updates.
> 
> I am currently working on simulating a low-energy injection beamline (from a DC Gun into a Linac) as well as a diagnostic setup for a DC Gun. In both beamlines, an alpha magnet plays a central role, and I would like to simulate it accurately.
> 
> The combination of low energies (~50 keV) and high currents (up to 2 A) makes it necessary to simulate space charge forces. Simulating the interaction of space charge forces with alpha magnets is quite rare. Currently, I am using ASTRA from DESY, but it requires extra work with custom field maps and coordinate transformations - which I am still working on.
> 
> ImpactX would offer a fast and modern codebase and open up many possibilities, especially for optimizing the magnet design for both beamlines.
> 
> I would be happy to assist with implementation or testing. While I have experience with using particle tracking codes, I am not deeply familiar with their inner workings.
> 
> TL;DR: It would be great to have alpha magnet support in ImpactX. Alternatively, guidance on simulating such a scenario using a combination of ImpactX and WarpX - as the latter was mentioned in [#370](https://github.com/BLAST-ImpactX/impactx/issues/370) - would be highly appreciated.

**Comment by ax3l** [comment]
Hi @KroneBonn,

Thank you for reaching out, the kind words, and explaining your need.

On our end, Alpha Magnets are currently not on any of our funded DOE HEP/BES/NP roadmaps, but we would be happy to collaborate to apply for funding with you, do sub-contracts, etc.. Thus, we cannot wing it on the side, unfortunately.

What is totally possible is to [implement this new element type](https://impactx.readthedocs.io/en/latest/usage/howto/add_element.html) as a contribution from you, and we review, merge and maintain the functionality in the mainline of ImpactX for everyone.

If you want to go fast, here is an update since your initial request:
We since added a [LinearMap](https://impactx.readthedocs.io/en/latest/usage/python.html#impactx.elements.LinearMap) element ([detailed example](https://impactx.readthedocs.io/en/latest/usage/examples/linear_map/README.html)). Looking at the [TraceWin documentation](https://irfu.cea.fr/dacm/logiciels/codesdacm/tracewin/tracewin.pdf) above, they implement Alpha Magnets as linear map. If that model has sufficient fidelity for you, you could calculate the matrix entries and put it into the ImpactX `LinearMap` element.

Additionally, if that model is useful you could if you like add it as an explicit linear element, like we also add linear Quad and Drift models.

Would any of those be a way to go?

**Comment by ax3l** [comment]
@KroneBonn is the linear map a way to help you on the short-term do your alpha magnet modeling?

**Comment by KroneBonn** [comment]
@ax3l Sorry. I totally forget to answer. Thanks for the clarification and explanation.

Actually the implementation as linear element looks interesting. Unfortunately I was not yet able to have a deeper look into it.

As soon as I am able to put a little work into it I am happy to contribute it here :)
